File: backend/app/schemas/equipment.py
"""Equipment Pydantic şemaları — Pydantic v2.

Tüm input şemaları extra="forbid" kullanır.
Status değerleri: aktif | bakimda | hasarli | hizmetdisi
"""
import uuid
from datetime import date, datetime
from decimal import Decimal
from typing import List, Literal, Optional

from pydantic import BaseModel, ConfigDict, Field

EQUIPMENT_STATUS = Literal["aktif", "bakimda", "hasarli", "hizmetdisi"]


# ── EquipmentCreate ───────────────────────────────────────────────────────────

class EquipmentCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    name: str
    equipment_type: Optional[str] = None
    serial_no: Optional[str] = None
    brand: Optional[str] = None
    model: Optional[str] = None
    purchase_date: Optional[date] = None
    purchase_cost: Optional[Decimal] = Field(default=None, ge=0)
    status: EQUIPMENT_STATUS = "aktif"
    assigned_person_id: Optional[uuid.UUID] = None
    last_maintenance_date: Optional[date] = None
    next_maintenance_date: Optional[date] = None
    insurance_expiry_date: Optional[date] = None
    notes: Optional[str] = None
    is_active: bool = True


# ── EquipmentUpdate ───────────────────────────────────────────────────────────

class EquipmentUpdate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    name: Optional[str] = None
    equipment_type: Optional[str] = None
    serial_no: Optional[str] = None
    brand: Optional[str] = None
    model: Optional[str] = None
    purchase_date: Optional[date] = None
    purchase_cost: Optional[Decimal] = Field(default=None, ge=0)
    status: Optional[EQUIPMENT_STATUS] = None
    assigned_person_id: Optional[uuid.UUID] = None
    last_maintenance_date: Optional[date] = None
    next_maintenance_date: Optional[date] = None
    insurance_expiry_date: Optional[date] = None
    notes: Optional[str] = None
    is_active: Optional[bool] = None


# ── EquipmentOut ──────────────────────────────────────────────────────────────

class EquipmentOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    club_id: uuid.UUID
    name: str
    equipment_type: Optional[str] = None
    serial_no: Optional[str] = None
    brand: Optional[str] = None
    model: Optional[str] = None
    purchase_date: Optional[date] = None
    purchase_cost: Optional[Decimal] = None
    status: str
    assigned_person_id: Optional[uuid.UUID] = None
    assigned_person_name: Optional[str] = None   # computed — router'da doldurulur
    last_maintenance_date: Optional[date] = None
    next_maintenance_date: Optional[date] = None
    insurance_expiry_date: Optional[date] = None
    notes: Optional[str] = None
    is_active: bool
    is_deleted: bool
    created_at: datetime
    updated_at: datetime


class EquipmentListOut(BaseModel):
    items: List[EquipmentOut]
    total: int
    skip: int
    limit: int


# ── MaintenanceDueOut — GET /equipment/maintenance-due ────────────────────────

class MaintenanceDueOut(EquipmentOut):
    """Bakım/sigorta uyarısı hesaplanmış alanlarla."""
    maintenance_due: bool = False
    maintenance_days_remaining: Optional[int] = None  # negatif = gecikti
    insurance_due: bool = False
    insurance_days_remaining: Optional[int] = None    # negatif = gecikti


# ── MaintenanceRecordCreate ───────────────────────────────────────────────────

class MaintenanceRecordCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    maintenance_date: date
    maintenance_type: Optional[str] = None
    description: Optional[str] = None
    cost: Optional[Decimal] = Field(default=None, ge=0)
    performed_by: Optional[str] = None
    next_maintenance_date: Optional[date] = None
    notes: Optional[str] = None


# ── MaintenanceRecordUpdate ───────────────────────────────────────────────────

class MaintenanceRecordUpdate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    maintenance_date: Optional[date] = None
    maintenance_type: Optional[str] = None
    description: Optional[str] = None
    cost: Optional[Decimal] = Field(default=None, ge=0)
    performed_by: Optional[str] = None
    next_maintenance_date: Optional[date] = None
    notes: Optional[str] = None


# ── MaintenanceRecordOut ──────────────────────────────────────────────────────

class MaintenanceRecordOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    club_id: uuid.UUID
    equipment_id: uuid.UUID
    maintenance_date: date
    maintenance_type: Optional[str] = None
    description: Optional[str] = None
    cost: Optional[Decimal] = None
    performed_by: Optional[str] = None
    next_maintenance_date: Optional[date] = None
    notes: Optional[str] = None
    recorded_by_user_id: Optional[uuid.UUID] = None
    is_deleted: bool
    created_at: datetime
    updated_at: datetime


class MaintenanceRecordListOut(BaseModel):
    items: List[MaintenanceRecordOut]
    total: int

File: frontend/src/types/equipment.ts
export type EquipmentStatus = 'aktif' | 'bakimda' | 'hasarli' | 'hizmetdisi'

export interface Equipment {
  id: string
  club_id: string
  name: string
  equipment_type: string | null
  serial_no: string | null
  brand: string | null
  model: string | null
  purchase_date: string | null        // ISO date "YYYY-MM-DD"
  purchase_cost: string | null        // Decimal as string from API
  status: EquipmentStatus
  assigned_person_id: string | null
  assigned_person_name: string | null
  last_maintenance_date: string | null
  next_maintenance_date: string | null
  insurance_expiry_date: string | null
  notes: string | null
  is_active: boolean
  is_deleted: boolean
  created_at: string
  updated_at: string
}

export interface EquipmentListResponse {
  items: Equipment[]
  total: number
  skip: number
  limit: number
}

export interface EquipmentCreate {
  name: string
  equipment_type?: string | null
  serial_no?: string | null
  brand?: string | null
  model?: string | null
  purchase_date?: string | null
  purchase_cost?: number | null
  status?: EquipmentStatus
  assigned_person_id?: string | null
  last_maintenance_date?: string | null
  next_maintenance_date?: string | null
  insurance_expiry_date?: string | null
  notes?: string | null
  is_active?: boolean
}

export interface EquipmentUpdate {
  name?: string
  equipment_type?: string | null
  serial_no?: string | null
  brand?: string | null
  model?: string | null
  purchase_date?: string | null
  purchase_cost?: number | null
  status?: EquipmentStatus | null
  assigned_person_id?: string | null
  last_maintenance_date?: string | null
  next_maintenance_date?: string | null
  insurance_expiry_date?: string | null
  notes?: string | null
  is_active?: boolean
}

// maintenance-due endpoint ek alanları
export interface MaintenanceDueEquipment extends Equipment {
  maintenance_due: boolean
  maintenance_days_remaining: number | null
  insurance_due: boolean
  insurance_days_remaining: number | null
}

export interface MaintenanceDueListResponse {
  items: MaintenanceDueEquipment[]
  total: number
}

export interface MaintenanceRecord {
  id: string
  club_id: string
  equipment_id: string
  maintenance_date: string
  maintenance_type: string | null
  description: string | null
  cost: string | null
  performed_by: string | null
  next_maintenance_date: string | null
  notes: string | null
  recorded_by_user_id: string | null
  is_deleted: boolean
  created_at: string
  updated_at: string
}

export interface MaintenanceRecordListResponse {
  items: MaintenanceRecord[]
  total: number
}

export interface MaintenanceRecordCreate {
  maintenance_date: string
  maintenance_type?: string | null
  description?: string | null
  cost?: number | null
  performed_by?: string | null
  next_maintenance_date?: string | null
  notes?: string | null
}

export interface MaintenanceRecordUpdate {
  maintenance_date?: string
  maintenance_type?: string | null
  description?: string | null
  cost?: number | null
  performed_by?: string | null
  next_maintenance_date?: string | null
  notes?: string | null
}
